README.md
redis-em-mutex
==============

Author: Rafał Michalski  (mailto:rafal@yeondir.com)

* http://github.com/royaltm/redis-em-mutex

DESCRIPTION
-----------

__redis-em-mutex__ is the cross server/process/fiber|owner EventMachine + Redis based semaphore.

FEATURES
--------

* EventMachine reactor based
* carefully designed, well thought out locking pattern
  (NOT the flawed SETNX/GET/GETSET one from redis documentation page)
* no CPU-intensive sleep/polling while waiting for lock to become available;
  fibers waiting for the lock are signalled via Redis channel as soon as the lock
  is released (~< 1 ms)
* alternative fast "script" handler (server-side LUA script based - redis-server 2.6.x)
* multi-locks (all-or-nothing) locking (to prevent possible deadlocks when
  multiple semaphores are required to be locked at once)
* fiber-safe
* deadlock detection (only trivial cases: locking twice the same resource from the same owner)
* mandatory lock lifetime expiration (with refreshing)
* macro-style definitions (Mutex::Macro mixin)
* compatible with Synchrony::Thread::ConditionVariable
* extendable (beyond fibers) mutex ownership
* redis HA achievable with [redis-sentinel](http://redis.io/topics/sentinel) and [redis-sentinel](https://github.com/flyerhzm/redis-sentinel) gem.

BUGS/LIMITATIONS
----------------

* only for EventMachine
* NOT thread-safe (not meant to be)
* locking order between concurrent processes is undetermined (no FIFO between processes)
  however during {file:BENCHMARK.md BENCHMARKING} no starvation effect was observed.
* it's not nifty, rather somewhat complicated

REQUIREMENTS
------------

* ruby >= 1.9 (tested: ruby 1.9.3p374, 1.9.3-p194, 1.9.2-p320, 1.9.1-p378)
* http://github.com/redis/redis-rb ~> 3.0.2
* http://rubyeventmachine.com ~> 1.0.0
* (optional) http://github.com/igrigorik/em-synchrony
  But due to the redis/synchrony dependency em-synchrony will always be bundled
  and required.

INSTALL
-------

```
$ [sudo] gem install redis-em-mutex
```

#### Gemfile

```ruby
gem "redis-em-mutex", "~> 0.3.1"
```

#### Github

```
git clone git://github.com/royaltm/redis-em-mutex.git
```

UPGRADING
---------

0.2.x -> 0.3.x

To upgrade redis-em-mutex on production from 0.2.x to 0.3.x you must make sure the correct handler has been
selected. See more on HANDLERS below.

The "pure" and "script" handlers are incompatible. Two different handlers must not utilize the same semaphore-key space.

Because only the "pure" handler is compatible with redis-em-mutex <= 0.2.x, when upgrading live production make sure to add
`handler: :pure` option to `Redis::EM::Mutex.setup` or set the environment variable on production app servers:

```sh
  REDIS_EM_MUTEX_HANDLER=pure
  export REDIS_EM_MUTEX_HANDLER
```

Upgrading from "pure" to "script" handler requires that all "pure" handler locks __MUST BE DELETED__ from redis-server beforehand.
Neglecting that will result in possible deadlocks. The "script" handler assumes that the lock expiration process is handled
by redis-server's PEXPIREAT feature. The "pure" handler does not set timeouts on keys. It handles expiration differently.

USAGE
-----

```ruby
    require 'em-synchrony'
    require 'redis-em-mutex'

    Redis::EM::Mutex.setup(size: 10, url: 'redis:///1', expire: 600)

    # or

    Redis::EM::Mutex.setup do |opts|
      opts.size = 10
      opts.url = 'redis:///1'
    end


    EM.synchrony do
      Redis::EM::Mutex.synchronize('resource.lock') do
        # ... do something with resource
      end

      # or 

      mutex = Redis::EM::Mutex.new('resource.lock')
      mutex.synchronize do
        # ... do something with resource
      end

      # or

      begin
        mutex.lock
        # ... do something with resource
      ensure
        mutex.unlock
      end

      # ...

      Redis::EM::Mutex.stop_watcher
      EM.stop
    end
```

### Handlers

There are 2 different mutex implementations since version 0.3.

* The "pure" classic handler utilizes redis optimistic transaction commands (watch/multi).
  This handler works with redis-server 2.4.x and later.
* The new "script" handler takes advantage of fast atomic server side operations written in LUA.
  Therefore the "script" handler is compatible only with redis-server 2.6.x and later.

__IMPORTANT__: The "pure" and "script" implementations are incompatible. The values that each handler stores in semaphore keys have different meaning to them.
You can not operate on the same set of keys using both handlers from e.g. different applications or application versions.
See UPGRADING for more info on this.

You choose your preferred implementation with `handler` option:

```ruby
    Redis::EM::Mutex.setup(handler: :script)
    Redis::EM::Mutex.handler # "Redis::EM::Mutex::ScriptHandlerMixin"

    # or

    Redis::EM::Mutex.setup do |opts|
      opts.handler = :pure
    end
    Redis::EM::Mutex.handler # "Redis::EM::Mutex::PureHandlerMixin"
```

You may also setup `REDIS_EM_MUTEX_HANDLER` environment variable to preferred implementation name.
Passing `handler` option to {Redis::EM::Mutex.setup} overrides environment variable.

The default handler option is `auto` which selects best handler available for your redis-server.
It's good for quick sandbox setup, however you should set explicitly which handler you require on production.

The differences:

* Performance. The "script" handler is faster then the "pure" handler.
  The "pure" handler generates twice as much CPU load as "script" handler.
  See {file:BENCHMARK.md BENCHMARK}.

* Expiration. The "script" implementation handler uses PEXPIREAT to mark semaphore life-time.
  The "pure" handler stores semaphore expiry timestamp in key value.
  Therefore the "script" handler can't refresh semaphores once they expire.
  The "pure" handler on the other hand could refresh expired semaphore but only
  if nothing else has locked on that expired key.

To detect feature of the current handler:

```ruby
  Redis::EM::Mutex.can_refresh_expired?           # true / false
  Redis::EM::Mutex.new(:foo).can_refresh_expired? # true / false
```

### Namespaces

```ruby
  Redis::EM::Mutex.setup(ns: 'Tudor')

  # or multiple namespaces:

  ns = Redis::EM::Mutex::NS.new('Tudor')

  EM.synchrony do
    ns.synchronize('Boscogne') do
      # .... do something special with Tudor:Boscogne
    end

    EM.stop
  end
```

### Multi-locks

This feature enables you to lock more then one key at the same time.
The multi key semaphores are deadlock-safe.

The classic deadlock example scenario with multiple resources:

* A acquires lock on resource :foo
* B acquires lock on resource :bar
* A tries to lock on resource :bar still keeping the :foo
* but at the same time B tries to acquire :foo while keeping the :bar.
* The deadlock occurs.

```ruby
  EM.synchrony do
    Redis::EM::Mutex.synchronize('foo', 'bar', 'baz') do
      # .... do something special with foo, bar and baz
    end

    # ...
    EM.stop
  end
```

### Locking options

```ruby
  EM.synchrony do
    begin
      Redis::EM::Mutex.synchronize('foo', 'bar', block: 0.25) do
        # .... do something with foo and bar
      end
    rescue Redis::EM::Mutex::MutexTimeout
      # ... locking timed out
    end

    Redis::EM::Mutex.synchronize('foo', 'bar', expire: 60) do |mutex|
      # .... do something with foo and bar in less than 60 seconds
      if mutex.refresh(120)
        # now we have additional 120 seconds until lock expires
      else
        # too late
      end
    end

    # ...
    EM.stop
  end
```

### Macro-style definition

Idea of macro-style definition was borrowed from http://github.com/kenn/redis-mutex.
Redis::EM::Mutex::Macro is a mixin which protects selected instance methods of a class with a mutex.
The locking scope will be Mutex global namespace + class name + method name.

```ruby
  class TheClass
    include Redis::EM::Mutex::Macro
  
    auto_mutex
    def critical_run
      # ... do some critical stuff
      # ....only one fiber in one process on one machine is executing
      #     this instance method of any instance of defined class
    end

    auto_mutex expire: 100, ns: '**TheClass**'
    # all critical methods defined later will inherit above options unless overridden
  
    auto_mutex # start and stop will be protected
    def start
      # ...
    end

    def stop
      # ...
    end

    no_auto_mutex
    def some_unprotected_method
      # ...
    end

    auto_mutex :run_long, expire: 100000, block: 10, on_timeout: :cant_run
    def run_long
      # ...
    end

    def cant_run
      # ...
    end

    def foo
      # ...
    end
    auto_mutex :foo, block: 0, on_timeout: proc { puts 'bar!' }

  end
```

### ConditionVariable

`Redis::EM::Mutex` may be used with `EventMachine::Synchrony::Thread::ConditionVariable`
in place of `EventMachine::Synchrony::Thread::Mutex`.

```ruby
  mutex = Redis::EM::Mutex.new('resource')
  resource = EM::Synchrony::Thread::ConditionVariable.new
  EM::Synchrony.next_tick do
    mutex.synchronize {
      # current fiber now needs the resource
      resource.wait(mutex)
      # current fiber can now have the resource
    }
  end
  EM::Synchrony.next_tick do
    mutex.synchronize {
      # current fiber has finished using the resource
      resource.signal
    }
  end
```

### Advanced

#### Customized owner

In some cases you might want to extend the ownership of a locked semaphore beyond the current fiber.
That is to be able to "share" a locked semaphore between group of fibers associated with some external resource.

For example we want to allow only one connection from one ip address to be made to our group of servers
at the same time.
Of course you could configure some kind of load balancer that distributes your load to do that for you,
but let's just assume that the load balancer has no such features available.

The default "owner" of a locked semaphore consists of 3 parts: "Machine uuid" + "process pid" + "fiber id".
Redis::EM::Mutex allows you to replace the last part of the "owner" with your own value.
To customize mutex's "ownership" you should add :owner option to the created mutex,
passing some well defined owner id.
In our case the "scope" of a mutex will be the client's ip address and the "owner id"
will be an __id__ of an EventMachine::Connection object created upon connection from the client.
This way we make sure that any other connection from the same ip address made to any server and process
will have different ownership defined. At the same time all the fibers working with that Connection instance
can access mutex's #refresh and #unlock methods concurrently.

```ruby
  module MutexedConnection
    def post_init
      @buffer = ""
      @mutex = nil
      ip = Socket.unpack_sockaddr_in(get_peername).last
      Fiber.new do
        @mutex = Redis::EM::Mutex.lock(ip, block: 0, owner: self.__id__)
        if @mutex
          send_data "Hello #{ip}!"
          #... initialize stuff
        else
          send_data "Duplicate connection not allowed!"
          close_connection_after_writing
        end
      end.resume
    end

    def unbind
      if @mutex
        Fiber.new { @mutex.unlock! }.resume
      end
    end

    def receive_data(data)
      @buffer << data
      return unless @mutex
      Fiber.new do
        if @mutex.refresh # refresh expiration timeout
          # ... handle data
        else
          close_connection_after_writing
        end
      end.resume
    end
  end

  EM::Synchrony.run {
    Redis::EM::Mutex.setup(size: 10, url: 'redis:///1', expire: 600)
    EM.start_server "0.0.0.0", 6969, MutexedConnection
  }
```

#### Forking live

You may safely fork process while running event reactor and having locked semaphores.
The locked semaphores in newly forked process will become unlocked while
their locked status in parent process will be preserved.


```ruby
  mutex = Redis::EM::Mutex.new('pirkaff', 'roshinu', expire: 60)

  EM.synchrony do
    mutex.lock

    EM.fork_reactor do
      Fiber.new do
        mutex.locked? # true
        mutex.owned?  # false
        mutex.synchronize do
          mutex.locked? # true
          mutex.owned?  # true

          # ...
        end

        # ...
        Redis::EM::Mutex.stop_watcher
        EM.stop
      end.resume
    end

    mutex.locked? # true
    mutex.owned?  # true

    mutex.unlock
    mutex.owned?  # false

    # ...
    Redis::EM::Mutex.stop_watcher
    EM.stop
  end
```

#### Redis factory

Want to use some non-standard redis options or customized redis client?
`redis_factory` option to the rescue.

High Availability example setup with redis-sentinel:

```ruby
  gem 'redis-sentinel', '~> 1.1.4'
  require 'redis-em-mutex'
  require 'redis-sentinel'
  Redis::Client.class_eval do
    define_method(:sleep) {|n| EM::Synchrony.sleep(n) }
  end

  REDIS_OPTS = {password: 'fight or die', db: 1}
  SENTINEL_OPTS = {
    master_name: "femto",
    sentinels: [{host: "wyald", port: 26379}, {host: "zodd", port: 26379}],
    failover_reconnect_timeout: 30
  }

  Redis::EM::Mutex.setup(REDIS_OPTS) do |config|
    config.size = 5                 # redis pool size
    config.reconnect_max = :forever # reconnect watcher forever
    config.redis_factory = proc do |opts|
      Redis.new opts.merge SENTINEL_OPTS
    end
  end
```

ADVOCACY
--------

Interesting (not eventmachine oriented) ruby-redis-mutex implementations:

* [mlanett/redis-lock](https://github.com/mlanett/redis-lock)
  Robust, well thought out and nice to use as it simply adds lock/unlock
  commands to Redis.
  Similar concept of locking/unlocking pattern (compared to the "pure" handler)
  though it uses two redis keys for keeping owner and lifetime expiration separately.
  "pure" handler stores both in one key, so less redis operations are involved.
  Blocked lock failure is handled by sleep/polling which involves more cpu load
  on ruby and redis. You may actually see it by running `time test/stress.rb`
  tool on both implementations and compare user/sys load.

* [dv/redis-semaphore](https://github.com/dv/redis-semaphore)
  Very promising experiment. Utilizes BLPOP to provide real FIFO queue of
  lock acquiring processes. In this way it doesn't need polling nor other means
  of signaling that the lock is available to those in waiting queue.
  This one could be used with EM straight out without any patching.

  IMHO the solution has two drawbacks:

  - no lifetime expiration or other means of protection from failure of a lock owner process;
    still they are trying hard to implement it [now](https://github.com/dv/redis-semaphore/pull/5)
    and I hope they will succeed.

  - the redis keys are used in an inversed manner: the lack of a key means that the lock is gone.
    On the contrary, when the lock is being released, the key is created and kept.
    This is not a problem when you have some static set of keys. However it might be a problem
    when you need to use lock keys based on random resources and you would need to implement
    some garbage collector to prevent redis from eating to much memory.


HISTORY.md
0.3.1
- fixed: script handler multi-lock time to wait on failure
- fixed: bench script randomizes subsets of multi key sets

0.3.0
- fixed: optimized pure handler
- added redis/em-connection-pool no more em-synchrony/connection_pool dependency
- added redis gem dependency updated to 3.0.2
- added owner_ident specs
- added redis-lua script handler
- added #can_refresh_expired? handler feature detection
- added handlers

0.2.3
- fixed: rare unlock! race condition introduced in 0.2.1
  manifesting as deadlock exception when no deadlock should occur

0.2.2
- fixed: uuid must be set only once
- fixed: forking process or setup before reactor is running would hang
  current fiber forever on first failed lock
- added: new setup option :redis_factory
- fixed: doc format errors

0.2.1
- fixed: sleep
- fixed: possible deadlock after #unlock! with multiple names

0.2.0
- added compatibility with EM::Synchrony::Thread::ConditionVariable
- added #sleep and #wakeup
- moved Redis::EM::Mutex::NS to autoloaded file
- added #unlock! and now mutex object stores ownership information
- added #expired? and other expiration status methods
- added watching? helper
- added customizable ownership

0.1.2
- features macro-style definitions
- fixed: setup with :redis
- added macros spec
- added more setup features spec

0.1.1
- fixed: namespaces didn't work
- added namespaces spec
- fixed: rare race condition could raise ArgumentError inside Mutex#lock()
- fixed: Mutex.stop_watching raised MutexError after changing internal state
- fixed: semaphores spec updated
- fixed: ruby-1.9.1 has no SecureRandom.uuid method

0.1.0
- first release


lib/redis/em-mutex.rb
# -*- coding: UTF-8 -*-
require 'ostruct'
require 'securerandom'
require 'redis/connection/synchrony' unless defined? Redis::Connection::Synchrony
require 'redis'

class Redis
  module EM
    # Cross machine-process-fiber EventMachine + Redis based semaphore.
    #
    # WARNING:
    #
    # Methods of this class are NOT thread-safe.
    # They are machine/process/fiber-safe.
    # All method calls must be invoked only from EventMachine's reactor thread.
    # Wrap mutex calls in EventMachine.shedule from non-reactor threads.
    #
    # - The terms "lock" and "semaphore" used in documentation are synonims.
    # - The term "owner" denotes a Ruby Fiber executing code in the scope of Machine/Process/Fiber
    #   possessing exclusively a named semaphore(s).
    #
    class Mutex

      autoload :NS, 'redis/em-mutex/ns'
      autoload :Macro, 'redis/em-mutex/macro'

      module Errors
        class MutexError < RuntimeError; end
        class MutexTimeout < MutexError; end
      end

      include Errors
      extend Errors

      SIGNAL_QUEUE_CHANNEL = "::#{self.name}::"
      AUTO_NAME_SEED = '__@'
      DEFAULT_RECONNECT_MAX_RETRIES = 10
      @@default_expire = 3600*24
      @@name_index = AUTO_NAME_SEED
      @@redis_pool = nil
      @@watching = false
      @@signal_queue = Hash.new {|h,k| h[k] = []}
      @@ns = nil
      @@handler = nil
      @@uuid ||= if SecureRandom.respond_to?(:uuid)
        SecureRandom.uuid
      else
        SecureRandom.base64(24)
      end

      private

      def signal_queue; @@signal_queue end
      def uuid; @@uuid; end
      def redis_pool; @@redis_pool end

      public

      # Selected implementation handler module name
      def self.handler; @@handler && @@handler.name end

      # Whether selected implementation handler
      # supports refreshing of already expired locks.
      def self.can_refresh_expired?
        @@handler.can_refresh_expired?
      end

      # Whether selected implementation handler
      # supports refreshing of already expired locks.
      def can_refresh_expired?
        @@handler.can_refresh_expired?
      end

      # Creates a new cross machine/process/fiber semaphore
      #
      #   Redis::EM::Mutex.new(*names, options = {})
      #
      # - *names = lock identifiers - if none they are auto generated
      # - options = hash:
      # - :name - same as *names (in case *names arguments were omitted)
      # - :block - default block timeout
      # - :expire - default expire timeout (see: Mutex#lock and Mutex#try_lock)
      # - :ns - local namespace (otherwise global namespace is used)
      # - :owner - owner definition instead of Fiber#__id__
      #
      # Raises MutexError if used before Mutex.setup.
      # Raises ArgumentError on invalid options.
      def initialize(*args)
        raise MutexError, "call #{self.class}::setup first" unless @@redis_pool
        self.class.setup_handler unless @@handler

        opts = args.last.kind_of?(Hash) ? args.pop : {}

        @names = args
        @names = Array(opts[:name] || "#{@@name_index.succ!}.lock") if @names.empty?
        @slept = {}
        raise ArgumentError, "semaphore names must not be empty" if @names.empty?
        @multi = !@names.one?
        @ns = opts[:ns] || @@ns
        @ns_names = @ns ? @names.map {|n| "#@ns:#{n}".freeze }.freeze : @names.map {|n| n.to_s.dup.freeze }.freeze
        @marsh_names = Marshal.dump(@ns_names)
        self.expire_timeout = opts[:expire] if opts.key?(:expire)
        self.block_timeout = opts[:block] if opts.key?(:block)
        self.extend(@@handler)
        post_init(opts)
      end

      attr_reader :names, :ns, :block_timeout
      alias_method :namespace, :ns

      def expire_timeout; @expire_timeout || @@default_expire; end

      def expire_timeout=(value)
        raise ArgumentError, "#{self.class.name}\#expire_timeout value must be greater than 0" unless (value = value.to_f) > 0
        @expire_timeout = value
      end

      def block_timeout=(value)
        @block_timeout = value.nil? ? nil : value.to_f
      end

      # Releases the lock unconditionally.
      # If the semaphore wasn't locked by the current owner it is silently ignored.
      # Returns self.
      def unlock
        unlock!
        self
      end

      # Wakes up currently sleeping fiber on a mutex.
      def wakeup(fiber)
        fiber.resume if @slept.delete(fiber)
      end

      # for compatibility with EventMachine::Synchrony::Thread::ConditionVariable
      alias_method :_wakeup, :wakeup

      # Releases the lock and sleeps `timeout` seconds if it is given and non-nil or forever.
      # Raises MutexError if mutex wasn't locked by the current owner.
      # Raises MutexTimeout if #block_timeout= was set and timeout
      # occured while locking after sleep.
      # If code block is provided it is executed after waking up, just before grabbing a lock.
      def sleep(timeout = nil)
        raise MutexError, "can't sleep #{self.class} wasn't locked" unless unlock!
        start = Time.now
        current = Fiber.current
        @slept[current] = true
        if timeout
          timer = ::EM.add_timer(timeout) do
            wakeup(current)
          end
          Fiber.yield
          ::EM.cancel_timer timer
        else
          Fiber.yield
        end
        @slept.delete current
        yield if block_given?
        raise MutexTimeout unless lock
        Time.now - start
      end

      # Execute block of code protected with semaphore.
      # Code block receives mutex object.
      # Returns result of code block.
      #
      # If `block_timeout` or Mutex#block_timeout is set and
      # lock isn't obtained within `block_timeout` seconds this method raises
      # MutexTimeout.
      def synchronize(block_timeout=nil)
        if lock(block_timeout)
          begin
            yield self
          ensure
            unlock
          end
        else
          raise MutexTimeout
        end
      end

      # Returns true if watcher is connected
      def watching?; @@watching == $$; end

      # Returns true if watcher is connected
      def self.watching?; @@watching == $$; end

      class << self
        attr_reader :reconnect_max_retries
        def reconnect_forever?
          @reconnect_max_retries < 0
        end
        def reconnect_max_retries=(max)
          @reconnect_max_retries = max == :forever ? -1 : max.to_i
        end
        def ns; @@ns; end
        def ns=(namespace); @@ns = namespace; end
        alias_method :namespace, :ns
        alias_method :namespace=, :ns=

        # Default value of expiration timeout in seconds.
        def default_expire; @@default_expire; end

        # Assigns default value of expiration timeout in seconds.
        # Must be > 0.
        def default_expire=(value)
          raise ArgumentError, "#{name}.default_expire value must be greater than 0" unless (value = value.to_f) > 0
          @@default_expire = value
        end

        # Setup redis database and other defaults.
        # MUST BE called once before any semaphore is created.
        #
        # opts = options Hash:
        #
        # global options:
        #
        # - :connection_pool_class - default is Redis::EM::ConnectionPool
        # - :redis_factory - default is proc {|redis_opts| Redis.new redis_opts }
        # - :handler - the default value is taken from envronment variable: REDIS_EM_MUTEX_HANDLER or :auto
        #     :pure   - optimistic locking commands based (redis-server >= 2.4)
        #     :script - server scripting based (redis-server >= 2.6)
        #     :auto   - autodetect and choose best available handler
        # - :expire   - sets global Mutex.default_expire 
        # - :ns       - sets global Mutex.namespace
        # - :reconnect_max - maximum num. of attempts to re-establish
        #   connection to redis server;
        #   default is 10; set to 0 to disable re-connecting;
        #   set to -1 or :forever to attempt forever
        #
        # redis connection options:
        #
        # - :size     - redis connection pool size
        #
        # passed directly to redis_factory:
        #
        # - :url      - redis server url
        #
        # or
        #
        # - :scheme   - "redis" or "unix"
        # - :host     - redis host
        # - :port     - redis port
        # - :password - redis password
        # - :db       - redis database number
        # - :path     - redis unix-socket path
        #
        # or
        #
        # - :redis    - initialized ConnectionPool of Redis clients.
        def setup(opts = {})
          stop_watcher
          @watcher_subscribed = nil
          opts = OpenStruct.new(opts)
          yield opts if block_given?
          redis_options = {:driver => :synchrony}
          redis_updater = proc do |redis|
            redis_options.update({
              :scheme => redis.scheme,
              :host   => redis.host,
              :port   => redis.port,
              :password => redis.password,
              :db       => redis.db,
              :path     => redis.path
            }.reject {|_k, v| v.nil?})
          end
          if (redis = opts.redis) && !opts.url
            redis_updater.call redis.client
          elsif opts.url
            redis_options[:url] = opts.url
          end
          redis_updater.call opts
          pool_size = (opts.size.to_i.nonzero? || 1).abs
          self.default_expire = opts.expire if opts.expire
          self.reconnect_max_retries = opts.reconnect_max if opts.reconnect_max
          @connection_pool_class = opts.connection_pool_class if opts.connection_pool_class.kind_of?(Class)
          @redis_options = redis_options
          @reconnect_max_retries ||= DEFAULT_RECONNECT_MAX_RETRIES
          @redis_factory = opts.redis_factory if opts.redis_factory
          @redis_factory ||= proc {|opts| Redis.new opts }
          raise TypeError, "redis_factory should respond to [] method" unless @redis_factory.respond_to?(:[])
          @@ns = opts.ns if opts.ns
          unless (@@redis_pool = redis)
            unless @connection_pool_class
              begin
                require 'redis/em-connection-pool' unless defined?(Redis::EM::ConnectionPool)
              rescue LoadError
                raise ":connection_pool_class required; could not fall back to Redis::EM::ConnectionPool"
              end
              @connection_pool_class = Redis::EM::ConnectionPool
            end
            @@redis_pool = @connection_pool_class.new(size: pool_size) do
              @redis_factory[redis_options]
            end
          end
          @redis_watcher = @redis_factory[redis_options]
          start_watcher if ::EM.reactor_running?

          case handler = opts.handler || @@handler
          when Module
            @@handler = handler
          when nil, Symbol, String
            setup_handler(handler)
          else
            raise TypeError, 'handler must be Symbol or Module'
          end
        end

        def setup_handler(handler = nil)
          handler = (handler || ENV['REDIS_EM_MUTEX_HANDLER'] || :auto).to_sym.downcase
          if handler == :auto
            return unless ::EM.reactor_running?
            handler = :script
            begin
              @@redis_pool.script(:exists)
            rescue Redis::CommandError
              handler = :pure
            end
          end
          const_name = "#{handler.to_s.capitalize}HandlerMixin"
          begin
            unless self.const_defined?(const_name)
              require "redis/em-mutex/#{handler}_handler"
            end
            @@handler = self.const_get(const_name)
          rescue LoadError, NameError
            raise "handler: #{handler} not found"
          end
        end

        def ready?
          !!@@redis_pool
        end

        # resets Mutex's automatic name generator
        def reset_autoname
          @@name_index = AUTO_NAME_SEED
        end

        def wakeup_queue_all
          @@signal_queue.each_value do |queue|
            queue.each {|h| h.call }
          end
        end

        # Initializes the "unlock" channel watcher. It's called by Mutex.setup
        # internally. Should not be used under normal circumstances.
        # If EventMachine is to be re-started (or after EM.fork_reactor) this method may be used instead of
        # Mutex.setup for "lightweight" startup procedure.
        def start_watcher
          raise MutexError, "call #{self.class}::setup first" unless @redis_watcher
          return if watching?
          if @@watching # Process id changed, we've been forked alive!
            @redis_watcher = @redis_factory[@redis_options]
            @@signal_queue.clear
          end
          @@watching = $$
          retries = 0
          Fiber.new do
            begin
              @redis_watcher.subscribe(SIGNAL_QUEUE_CHANNEL) do |on|
                on.subscribe do |channel,|
                  if channel == SIGNAL_QUEUE_CHANNEL
                    @watcher_subscribed = true
                    retries = 0
                    wakeup_queue_all
                  end
                end
                on.message do |channel, message|
                  if channel == SIGNAL_QUEUE_CHANNEL
                    sig_match = {}
                    Marshal.load(message).each do |name|
                      sig_match[@@signal_queue[name].first] = true if @@signal_queue.key?(name)
                    end
                    sig_match.keys.each do |sig_proc|
                      sig_proc.call if sig_proc
                    end
                  end
                end
                on.unsubscribe do |channel,|
                  @watcher_subscribed = false if channel == SIGNAL_QUEUE_CHANNEL
                end
              end
              break
            rescue Redis::BaseConnectionError, EventMachine::ConnectionError => e
              @watcher_subscribed = false
              warn e.message
              retries+= 1
              if retries > reconnect_max_retries && reconnect_max_retries >= 0
                @@watching = false
              else
                sleep retries > 1 ? 1 : 0.1
              end
            end while watching?
          end.resume
          until @watcher_subscribed
            raise MutexError, "Can not establish watcher channel connection!" unless watching?
            fiber = Fiber.current
            ::EM.next_tick { fiber.resume }
            Fiber.yield
          end
        end

        # EM sleep helper
        def sleep(seconds)
          fiber = Fiber.current
          ::EM::Timer.new(seconds) { fiber.resume }
          Fiber.yield
        end

        # Stops the watcher of the "unlock" channel.
        # It should be called before stopping EvenMachine otherwise
        # EM might wait forever for channel connection to be closed.
        #
        # Raises MutexError if there are still some fibers waiting for lock.
        # Pass `true` to forcefully stop it. This might instead cause
        # MutexError to be raised in waiting fibers.
        def stop_watcher(force = false)
          return unless watching?
          raise MutexError, "call #{self.class}::setup first" unless @redis_watcher
          unless @@signal_queue.empty? || force
            raise MutexError, "can't stop: semaphores in queue"
          end
          @@watching = false
          if @watcher_subscribed
            @redis_watcher.unsubscribe SIGNAL_QUEUE_CHANNEL
            while @watcher_subscribed
              fiber = Fiber.current
              ::EM.next_tick { fiber.resume }
              Fiber.yield
            end
          end
        end

        # Remove all current Machine/Process locks.
        # Since there is no lock tracking mechanism, it might not be implemented easily.
        # If the need arises then it probably should be implemented.
        def sweep
          raise NotImplementedError
        end

        # Attempts to grab the lock and waits if it isn't available.
        # Raises MutexError if mutex was locked by the current owner
        # or if used before Mutex.setup.
        # Raises ArgumentError on invalid options.
        # Returns instance of Redis::EM::Mutex if lock was successfully obtained.
        # Returns `nil` if lock wasn't available within `:block` seconds.
        #
        #   Redis::EM::Mutex.lock(*names, options = {})
        #
        # - *names = lock identifiers - if none they are auto generated
        # - options = hash:
        # - :name - same as name (in case *names arguments were omitted)
        # - :block - block timeout
        # - :expire - expire timeout (see: Mutex#lock and Mutex#try_lock)
        # - :ns - namespace (otherwise global namespace is used)
        def lock(*args)
          mutex = new(*args)
          mutex if mutex.lock
        end

        # Execute block of code protected with named semaphore.
        # Returns result of code block.
        #
        #   Redis::EM::Mutex.synchronize(*names, options = {}, &block)
        # 
        # - *names = lock identifiers - if none they are auto generated
        # - options = hash:
        # - :name - same as name (in case *names arguments were omitted)
        # - :block - block timeout
        # - :expire - expire timeout (see: Mutex#lock and Mutex#try_lock)
        # - :ns - namespace (otherwise global namespace is used)
        # 
        # If `:block` is set and lock isn't obtained within `:block` seconds this method raises
        # MutexTimeout.
        # Raises MutexError if used before Mutex.setup.
        # Raises ArgumentError on invalid options.
        def synchronize(*args, &block)
          new(*args).synchronize(&block)
        end

      end

    end
  end
end

require 'redis/em-mutex/version'
